Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,605 @@
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#import <Starboard.h>
+#import <StubReturn.h>
+#import <Foundation/NSDecimal.h>
+#import <algorithm>
+
+#define IS_NEGATIVE(decimal) (decimal)->_isNegative
+
+#define IS_POSITIVE(decimal) !IS_NEGATIVE(decimal)
+
+#define IS_ZERO(decimal) (((decimal)->_length == 0) && IS_POSITIVE(decimal))
+
+#define INVALIDATE_DECIMAL(decimal) \
+    \
+decimal->_length = 0;               \
+    \
+decimal->_isNegative = YES;
+
+#define SET_TO_ZERO(decimal) \
+    \
+decimal->_length = 0;        \
+    \
+decimal->_isNegative = NO;   \
+    \
+NSDecimalCompact(decimal);
+
+#define VALIDATE_ARITHMETIC_INPUT(result, leftOperand, rightOperand)                 \
+    if (NSDecimalIsNotANumber(leftOperand) || NSDecimalIsNotANumber(rightOperand)) { \
+        INVALIDATE_DECIMAL(result);                                                  \
+        return NSCalculationOverflow;                                                \
+    }
+
+const int _MAX_EXP = 127;
+
+typedef struct {
+    NSDecimal decimal;
+    unsigned short* _mantissa;
+    unsigned short _remainder;
+    // only used for cases where we have carryover
+    unsigned int _extendedLen;
+} NSDecimalInternal;
+
+/**
+ @Status Interoperable
+*/
+void NSDecimalCopy(NSDecimal* destination, const NSDecimal* source) {
+    if (!source || !destination) {
+        return;
+    }
+
+    destination->_isNegative = source->_isNegative;
+    destination->_exponent = source->_exponent;
+    destination->_length = source->_length;
+    destination->_isCompact = source->_isCompact;
+    destination->_reserved = source->_reserved;
+    memcpy(&destination->_mantissa, &source->_mantissa, NSDecimalMaxSize * sizeof(unsigned short));
+}
+
+/**
+ @Status Interoperable
+*/
+BOOL NSDecimalIsNotANumber(const NSDecimal* dcm) {
+    THROW_NS_IF_FALSE(E_INVALIDARG, dcm != NULL);
+    return !dcm->_length && IS_NEGATIVE(dcm);
+}
+
+static void _InternalDecimalCopy(NSDecimal* destination, NSDecimalInternal* source) {
+    if (!source || !destination) {
+        return;
+    }
+    NSDecimalCopy(destination, &(source->decimal));
+    memcpy(destination->_mantissa, source->_mantissa, NSDecimalMaxSize * sizeof(unsigned short));
+}
+
+static NSCalculationError _subtraction(NSDecimal* result, const NSDecimal* left, const NSDecimal* right) {
+    // Note: All inputs have been normalized via NSDecimalNormalize
```

Comment:
Need you to mention more about your assumptions here - you have a lot more than just this
1) NSDecimalNormalize doesn't arrange the order of the two numbers
2) You also do other preprocessing so that subtraction here is safe. I was expecting the following subtraction
0001 FFFE
0000 FFFF
to fail, but it actually ended up succeeding because it ended up as
0000 FFFF
0000 0000
by the time it got here? Please qualify everything more x_____x

(also the above would probably be a good test case)